mage_ai/data_preparation/logging/logger_manager.py

```diff
@@ -79,36 +81,166 @@ def create_log_filepath_dir(self, path):
         if not os.path.exists(path):
             os.makedirs(path)
 
-    def get_log_filepath_prefix(self):
+    def get_log_filepath_prefix(self, include_date_hour_subpath=False):
         logs_dir = self.logs_dir or self.repo_config.variables_dir
+        current_date = datetime.utcnow().strftime('%Y%m%d')
+        current_hour = datetime.utcnow().strftime('%H')
 
         return os.path.join(
             logs_dir,
-            'pipelines',
+            PIPELINES_FOLDER,
             self.pipeline_uuid,
             LOGS_DIR,
             self.partition or '',
+            current_date if include_date_hour_subpath else '',
+            current_hour if include_date_hour_subpath else '',
         )
 
+    def get_log_filename(self):
+        filename = 'pipeline.log'
+        if self.block_uuid is not None:
+            filename = f'{self.block_uuid}.log'
+
+        return filename
+
+    def traverse_logs_dir(
+        self,
+        dir: str,
+        filename: str = None,
+        depth: int = 1,
+        start_timestamp: datetime = None,
+        end_timestamp: datetime = None,
+        file_ends_with: str = None,
+        write_date_depth: int = 1,  # Depth of folder with format YYYYMMDD
+    ):
+        """
+        Depending on which parent directory we start traversing from,
+        the write date YYYYMMDD folder will start at a certain depth.
+        """
+        date_depth = write_date_depth
+        hour_depth = date_depth + 1
+        start_date = None
+        start_hour = None
+        end_date = None
+        end_hour = None
+        if start_timestamp:
+            start_date = start_timestamp.strftime('%Y%m%d')
+            start_hour = start_timestamp.strftime('%H')
+        if end_timestamp:
+            end_date = end_timestamp.strftime('%Y%m%d')
+            end_hour = end_timestamp.strftime('%H')
+
+        subfolders_in_range, files_in_range = [], []
+        for f in os.scandir(dir):
+            if f.is_dir():
+                folder_timestamp = int(f.path.split('/')[-1]) if depth >= date_depth else None
+                write_date = int(f.path.split('/')[-2]) if depth == hour_depth else None
+                """
+                Compare YYYYMMDD and HH folder write dates/times with start/end
+                timestamps to see if we can skip checking certain folders.
+                """
+                if (start_timestamp is None and end_timestamp is None) or \
+                        (depth != date_depth and depth != hour_depth):
+                    subfolders_in_range.append(f.path)
+                elif start_timestamp is not None and end_timestamp is not None:
```

Comment:
Do you have duplicate code in this branch?

Reply:
I don't think so. There were a lot of conditions and combinations to check, so there are certain conditions that come up again but i thot it made the most sense organizing like this. I cleaned it up a bit and used some variables to make the conditions a little bit more succinct.